drivers/sensor/kx022/kx022.c

```diff
@@ -0,0 +1,523 @@
+		break;
+	case SENSOR_CHAN_ACCEL_XYZ:
+		kx022_sample_fetch_accel_xyz(dev);
+		break;
+	case SENSOR_CHAN_FREE_FALL:
+		kx022_motion_direction(dev);
+		break;
+	case SENSOR_CHAN_ALL:
+		kx022_sample_fetch_accel_xyz(dev);
+		kx022_tilt_pos(dev);
+		kx022_motion_direction(dev);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static inline void kx022_convert(struct sensor_value *val, int raw_val, float gain)
+{
+	int64_t dval;
+
+	/* Gain is in mg/LSB */
+	/* Convert to m/s^2 */
+	dval = ((int64_t)raw_val * gain * SENSOR_G) / 1000;
+	val->val1 = dval / 1000000LL;
+	val->val2 = dval % 1000000LL;
+}
+static inline void kx022_tilt_pos_get(struct sensor_value *val, int raw_val)
+{
+	uint16_t dval;
+
+	dval = (int16_t)raw_val;
+	val->val1 = dval;
+}
+
+static inline void kx022_motion_dir_get(struct sensor_value *val, int raw_val)
+{
+	uint16_t dval;
+
+	dval = (int16_t)raw_val;
+	val->val1 = dval;
+}
+static inline int kx022_get_channel(enum sensor_channel chan, struct sensor_value *val,
+				    struct kx022_data *data, float gain)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		kx022_convert(val, data->sample_x, gain);
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		kx022_convert(val, data->sample_y, gain);
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		kx022_convert(val, data->sample_z, gain);
+		break;
+	case SENSOR_CHAN_ACCEL_XYZ:
+		kx022_convert(val, data->sample_x, gain);
+		kx022_convert(val + 1, data->sample_y, gain);
+		kx022_convert(val + 2, data->sample_z, gain);
+		break;
+	case SENSOR_CHAN_FREE_FALL:
+		kx022_motion_dir_get(val, data->sample_motion_dir);
+		break;
+	case SENSOR_CHAN_NEAR_FAR:
+		kx022_tilt_pos_get(val, data->sample_tspp);
+		kx022_tilt_pos_get(val + 1, data->sample_tscp);
+		break;
+	case SENSOR_CHAN_ALL:
+		kx022_convert(val, data->sample_x, gain);
+		kx022_convert(val + 1, data->sample_y, gain);
+		kx022_convert(val + 2, data->sample_z, gain);
+		kx022_tilt_pos_get(val + 3, data->sample_tspp);
+		kx022_tilt_pos_get(val + 4, data->sample_tscp);
+		kx022_motion_dir_get(val + 5, data->sample_motion_dir);
+
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int kx022_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct kx022_data *data = dev->data;
+
+	return kx022_get_channel(chan, val, data, data->gain);
+}
+
+static const struct sensor_driver_api kx022_api_funcs = {
+	.attr_set = kx022_attr_set,
+#ifdef CONFIG_KX022_TRIGGER
+	.trigger_set = kx022_trigger_set,
+#endif
+	.sample_fetch = kx022_sample_fetch,
+	.channel_get = kx022_channel_get,
+};
+
+static int kx022_init(const struct device *dev)
+{
+	const struct kx022_config *const config = dev->config;
+	struct kx022_data *data = dev->data;
+	uint8_t chip_id;
+
+	data->comm_master = device_get_binding(config->comm_master_dev_name);
+	if (!data->comm_master) {
+		LOG_DBG("master not found: %s", config->comm_master_dev_name);
+		return -EINVAL;
+	}
+	k_msleep(2000);
```

Comment:
- this will delay reaching `main()` with at least 2s, is this really needed?
- constants need to be defined